File: core/src/main/scala/eu/ostrzyciel/jelly/core/internal/ProtoEncoderBase.scala
package eu.ostrzyciel.jelly.core.internal

import eu.ostrzyciel.jelly.core.*
import eu.ostrzyciel.jelly.core.proto.v1.*

/**
 * Base trait for Jelly proto encoders. Only for internal use.
 * @tparam TNode type of RDF nodes in the library
 */
private[core] trait ProtoEncoderBase[TNode, -TTriple, -TQuad]:
  protected val nodeEncoder: NodeEncoder[TNode]
  protected val converter: ProtoEncoderConverter[TNode, TTriple, TQuad]

  private val lastSubject: LastNodeHolder[TNode] = new LastNodeHolder()
  private val lastPredicate: LastNodeHolder[TNode] = new LastNodeHolder()
  private val lastObject: LastNodeHolder[TNode] = new LastNodeHolder()
  private var lastGraph: TNode | LastNodeHolder.NoValue.type = LastNodeHolder.NoValue

  private final def nodeToProtoWrapped(node: TNode, lastNodeHolder: LastNodeHolder[TNode]): SpoTerm =
    if node.equals(lastNodeHolder.node) then null
    else
      lastNodeHolder.node = node
      converter.nodeToProto(nodeEncoder, node)

  private final def graphNodeToProtoWrapped(node: TNode): GraphTerm =
    // Graph nodes may be null in Jena for example... so we need to handle that.
    if (node == null && lastGraph == null) || (node != null && node.equals(lastGraph)) then
      null
    else
      lastGraph = node
      converter.graphNodeToProto(nodeEncoder, node)

  protected final def tripleToProto(subject: TNode, predicate: TNode, `object`: TNode): RdfTriple =
    RdfTriple(
      subject = nodeToProtoWrapped(subject, lastSubject),
      predicate = nodeToProtoWrapped(predicate, lastPredicate),
      `object` = nodeToProtoWrapped(`object`, lastObject),
    )

  protected final def quadToProto(
    subject: TNode, predicate: TNode, `object`: TNode, graph: TNode
  ): RdfQuad = RdfQuad(
    subject = nodeToProtoWrapped(subject, lastSubject),
    predicate = nodeToProtoWrapped(predicate, lastPredicate),
    `object` = nodeToProtoWrapped(`object`, lastObject),
    graph = graphNodeToProtoWrapped(graph),
  )

  /**
   * Converts a triple to an RdfQuad object with a null graph.
   *
   * Used in RDF-Patch for triple add/delete operations.
   */
  protected final def tripleInQuadToProto(
    subject: TNode, predicate: TNode, `object`: TNode
  ): RdfQuad = RdfQuad(
    subject = nodeToProtoWrapped(subject, lastSubject),
    predicate = nodeToProtoWrapped(predicate, lastPredicate),
    `object` = nodeToProtoWrapped(`object`, lastObject),
    graph = null,
  )

File: core-java/src/main/java/eu/ostrzyciel/jelly/core/internal/ProtoEncoderBase.java
package eu.ostrzyciel.jelly.core.internal;

import eu.ostrzyciel.jelly.core.NodeEncoder;
import eu.ostrzyciel.jelly.core.ProtoEncoderConverter;
import eu.ostrzyciel.jelly.core.RdfTerm;
import eu.ostrzyciel.jelly.core.RowBufferAppender;
import eu.ostrzyciel.jelly.core.proto.v1.RdfStreamOptions;

/**
 * Base interface for Jelly proto encoders. Only for internal use.
 * @param <TNode> type of RDF nodes in the library
 */
public abstract class ProtoEncoderBase<TNode> implements RowBufferAppender {

    /**
     * RdfStreamOptions for this encoder.
     */
    protected final RdfStreamOptions options;
    protected final NodeEncoder<TNode> nodeEncoder;
    protected final ProtoEncoderConverter<TNode> converter;

    protected final LastNodeHolder<TNode> lastSubject = new LastNodeHolder<>();
    protected final LastNodeHolder<TNode> lastPredicate = new LastNodeHolder<>();
    protected final LastNodeHolder<TNode> lastObject = new LastNodeHolder<>();
    protected TNode lastGraph = null;

    protected ProtoEncoderBase(RdfStreamOptions options, ProtoEncoderConverter<TNode> converter) {
        this.options = options;
        this.nodeEncoder = NodeEncoderImpl.create(options, this);
        this.converter = converter;
    }

    protected final RdfTerm.Triple tripleToProto(TNode subject, TNode predicate, TNode object) {
        return new RdfTerm.Triple(
            nodeToProtoWrapped(subject, lastSubject),
            nodeToProtoWrapped(predicate, lastPredicate),
            nodeToProtoWrapped(object, lastObject)
        );
    }

    protected final RdfTerm.Quad quadToProto(TNode subject, TNode predicate, TNode object, TNode graph) {
        return new RdfTerm.Quad(
            nodeToProtoWrapped(subject, lastSubject),
            nodeToProtoWrapped(predicate, lastPredicate),
            nodeToProtoWrapped(object, lastObject),
            graphNodeToProtoWrapped(graph)
        );
    }

    /**
     * Converts a triple to an RdfQuad object with a null graph.
     * <p>
     * Used in RDF-Patch for triple add/delete operations.
     */
    protected final RdfTerm.Quad tripleInQuadToProto(TNode subject, TNode predicate, TNode object) {
        return new RdfTerm.Quad(
            nodeToProtoWrapped(subject, lastSubject),
            nodeToProtoWrapped(predicate, lastPredicate),
            nodeToProtoWrapped(object, lastObject),
            null
        );
    }

    private RdfTerm.SpoTerm nodeToProtoWrapped(TNode node, LastNodeHolder<TNode> lastNodeHolder) {
        if (node.equals(lastNodeHolder.node)) {
            return null;
        } else {
            lastNodeHolder.node = node;
            return converter.nodeToProto(nodeEncoder, node);
        }
    }

    private RdfTerm.GraphTerm graphNodeToProtoWrapped(TNode node) {
        // Graph nodes may be null in Jena for example... so we need to handle that.
        if ((node == null && lastGraph == null) || (node != null && node.equals(lastGraph))) {
            return null;
        } else {
            lastGraph = node;
            return converter.graphNodeToProto(nodeEncoder, node);
        }
    }
}
